panic: runtime error: invalid memory address or nil pointer dereference

# kapacitor RPM info:

Name        : kapacitor
Arch        : x86_64
Version     : 0.13.1
Release     : 1
Size        : 41 M
Repo        : installed
From repo   : /kapacitor-0.13.1.x86_64
Summary     : Time series data processing engine
URL         : github.com/influxdata/kapacitor
License     : MIT
Description : Time series data processing engine
# TICKscript

```
var ping_statistic = stream
        |from()
                .database('netbench')
                .retentionPolicy('default')
                .measurement('ping_statistic')
                .groupBy('hostName', 'isp', 'province')
        |window()
                .period(1m)
                .every(1m)

var avgLoss = ping_statistic
        |last('avgLoss')
                .as('avgLoss')

avgLoss
        |alert()
                .crit(lambda: "avgLoss" > 5)
                .log('/tmp/alerts_loss.log')
```
# kapacitor log

[httpd] 2016/09/06 23:38:04 I! 10.160.109.152 - - [06/Sep/2016:23:38:04 +0800] GET /kapacitor/v1/tasks/alert_netbench HTTP/1.1 200 5974 - curl/7.19.7 (x86_64-redhat-linux-gnu) libcurl/7.19.7 NSS/3.21 Basic ECC zlib/1.2.3 libidn/1.18 libssh2/1.4.2 e7153ecb-7447-11e6-8014-000000000000 6.812757ms
[alert_netbench:last3] 2016/09/06 23:39:03 E! failed to aggregate batch: field avgLoss missing from point cannot aggregate
# panic message:

panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x5f5375]

goroutine 184 [running]:
panic(0xca0ec0, 0xc820010100)
        /usr/local/go/src/runtime/panic.go:481 +0x3e6
github.com/influxdata/kapacitor/vendor/github.com/influxdata/influxdb/influxql.(_FloatFuncReducer).Emit(0xc8202d5490, 0x0, 0x0, 0x0)
        /root/go/src/github.com/influxdata/kapacitor/vendor/github.com/influxdata/influxdb/influxql/functions.gen.go:68 +0xa5
github.com/influxdata/kapacitor.(_floatPointEmitter).EmitPoint(0xc820b40510, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /root/go/src/github.com/influxdata/kapacitor/influxql.gen.go:166 +0x78
github.com/influxdata/kapacitor.(_InfluxQLNode).emit(0xc8200e5000, 0x7f3596286140, 0xc820b404e0, 0x0, 0x0)
        /root/go/src/github.com/influxdata/kapacitor/influxql.go:185 +0x90
github.com/influxdata/kapacitor.(_InfluxQLNode).runBatchInfluxQL(0xc8200e5000, 0x0, 0x0)
        /root/go/src/github.com/influxdata/kapacitor/influxql.go:162 +0x62c
github.com/influxdata/kapacitor.(_InfluxQLNode).runInfluxQLs(0xc8200e5000, 0x0, 0x0, 0x0, 0x0, 0x0)
        /root/go/src/github.com/influxdata/kapacitor/influxql.go:42 +0x83
github.com/influxdata/kapacitor.(_InfluxQLNode).(github.com/influxdata/kapacitor.runInfluxQLs)-fm(0x0, 0x0, 0x0, 0x0, 0x0)
        /root/go/src/github.com/influxdata/kapacitor/influxql.go:33 +0x4c
github.com/influxdata/kapacitor.(_node).start.func1(0xc8200e5000, 0x0, 0x0, 0x0)
        /root/go/src/github.com/influxdata/kapacitor/node.go:128 +0x7f
created by github.com/influxdata/kapacitor.(_node).start
        /root/go/src/github.com/influxdata/kapacitor/node.go:129 +0x53

@nathanielc , I have added debug code to  kapacitor:
```
diff --git a/influxql.gen.go b/influxql.gen.go
index c8720a3..7cd67f9 100644
--- a/influxql.gen.go
+++ b/influxql.gen.go
@@ -163,7 +163,17 @@ type floatPointEmitter struct {
        byName           bool
 }

-func (e *floatPointEmitter) EmitPoint() (models.Point, error) {
+func (e *floatPointEmitter) EmitPoint() (point models.Point, err error) {
+       defer func(){
+               if r := recover(); r != nil {
+                       fmt.Printf("\npanic: %v\n", r)
+                       fmt.Printf("e.baseReduceContext: %#v\n", e.baseReduceContext)
+                       fmt.Printf("\ne.emitter: %#v\n", e.emitter)
+
+                       point = models.Point{}
+                       err = ErrEmptyEmit
+               }
+       }()
```
When panic, the output is following:
```
panic: runtime error: invalid memory address or nil pointer dereference
e.baseReduceContext: kapacitor.baseReduceContext{as:"avgLoss", field:"avgLoss", name:"ping_statistic", group:"hostName=亦庄BGP,isp=联通,province=海南", dimensions:models.Dimensions{ByName:false, TagNames:[]string{"hostName", "isp", "province"}}, tags:models.Tags{"hostName":"亦庄BGP", "isp":"联通", "province":"海南"}, time:time.Time{sec:63609014914, nsec:106981616, loc:(*time.Location)(0xf9cf00)}, pointTimes:false, topBottomInfo:(*pipeline.TopBottomCallInfo)(nil)}

e.emitter: &influxql.FloatFuncReducer{prev:(*influxql.FloatPoint)(nil), fn:(influxql.FloatReduceFunc)(0x5c4c90)}
```
e.emitter.prev is nil, so in influxdb/influxql/functions.gen.go  line 72~75:

```
// Emit emits the point that was generated when reducing the points fed in with AggregateFloat.
func (r *FloatFuncReducer) Emit() []FloatPoint {
	return []FloatPoint{*r.prev}
}
```
executing *r.prev will panic.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -76,6 +76,7 @@ See the [API docs on technical preview](https://docs.influxdata.com/kapacitor/v1
 - [#1045](https://github.com/influxdata/kapacitor/issues/1045): Fix panic during replays.
 - [#1043](https://github.com/influxdata/kapacitor/issues/1043): logrotate.d ignores kapacitor configuration due to bad file mode
 - [#1100](https://github.com/influxdata/kapacitor/issues/1100): Fix issue with the Union node buffering more points than necessary.
+- [#872](https://github.com/influxdata/kapacitor/issues/872): Fix panic during failed aggregate results.
 
 ## v1.1.1 [2016-12-02]
 

--- a/influxql.go
+++ b/influxql.go
@@ -201,11 +201,13 @@ func (n *InfluxQLNode) runBatchInfluxQL() error {
 			}
 			n.timer.Resume()
 		} else {
-			if err := context.AggregateBatch(&b); err != nil {
+			err := context.AggregateBatch(&b)
+			if err == nil {
+				if err := n.emit(context); err != nil {
+					n.logger.Println("E! failed to emit batch:", err)
+				}
+			} else {
 				n.logger.Println("E! failed to aggregate batch:", err)
-			}
-			if err := n.emit(context); err != nil {
-				n.logger.Println("E! failed to emit batch:", err)
 			}
 		}
 		n.timer.Stop()

